libqtile/widget/decorations.py

```diff
@@ -0,0 +1,227 @@
+# Copyright (c) 2021, elParaguayo. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import math
+from typing import Any, List, Tuple
+
+from libqtile.log_utils import logger
+from libqtile.widget import base
+
+
+class _Decoration(base.PaddingMixin):
+    """
+        Base decoration class. Should not be called by
+        configs directly.
+    """
+
+    defaults = [
+        ("padding", 0, "Default padding")
+    ]  # type: List[Tuple[str, Any, str]]
+
+    def __init__(self, **config):
+        base.PaddingMixin.__init__(self, **config)
+        self.add_defaults(_Decoration.defaults)
+
+    @property
+    def height(self):
+        return self.parent.height
+
+    @property
+    def width(self):
+        return self.parent.width
+
+    @property
+    def drawer(self):
+        return self.parent.drawer
+
+    @property
+    def ctx(self):
+        return self.parent.drawer.ctx
+
+    def draw(self, parent):
+        self.parent = parent
+
+
+class RectDecoration(_Decoration):
+    """
+        Widget decoration that draws a rectangle behind the widget contents.
+
+        Only one colour can be set but decorations can be layered to achieve
+        multi-coloured effects.
+
+        Rectangles can be drawn as just the the outline or filled. Curved corners
+        can be obtained by setting the ``radius`` parameter.
+    """
+    defaults = [
+        ("filled", False, "Whether to fill shape"),
+        ("radius", 4, "Radius for corners (0 for square)"),
+        ("dec_colour", "#000000", "Colour for decoration"),
+        ("linewidth", 2, "Line width for decoration")
```

Comment:
`line_width`? or are you trying to be consistent with another place where there is `linewidth`?

Reply:
I think `drawer` and `Sep` both use `linewidth` but `graph` uses `line_width`. I'll change it as I think the latter is more readable.

Reply:
+1 for line_width where possible, hopefully no user-facing configuration options use linewidth!